External/Plugins/ASCompletion/Completion/ASComplete.cs

```diff
@@ -4439,6 +4439,58 @@ private static bool IsXmlType(ClassModel model)
                 && (model.QualifiedName == "XML" || model.QualifiedName == "XMLList");
         }
 
+        public static int ExpressionEndPosition(ScintillaControl sci, int position)
+        {
+            var member = ASContext.Context.CurrentMember;
+            var endPosition = member != null ? sci.PositionFromLine(member.LineTo) : sci.TextLength;
+            return ExpressionEndPosition(sci, position, endPosition);
+        }
+
+        public static int ExpressionEndPosition(ScintillaControl sci, int startPos, int endPos)
+        {
+            var characterClass = ScintillaControl.Configuration.GetLanguage(sci.ConfigurationLanguage).characterclass.Characters;
+            var groupCount = 0;
+            var spacesCount = 0;
+            var statementEnd = startPos;
+            var hadWS = false;
+            sci.Colourise(0, -1);
+            while (statementEnd < endPos)
+            {
+                if (sci.PositionIsOnComment(statementEnd) || sci.PositionIsInString(statementEnd))
+                {
+                    if (groupCount == 0) spacesCount++;
+                    statementEnd++;
+                    continue;
+                }
+                var c = (char) sci.CharAt(statementEnd++);
+                if (c == '(')
+                {
+                    groupCount++;
+                    hadWS = false;
+                    spacesCount = 0;
+                }
+                else if (c == ')')
+                {
+                    groupCount--;
+                    if (groupCount < 0) break;
+                }
+                else if (groupCount == 0)
+                {
+                    if (characterClass.Contains(c))
+                    {
+                        if (hadWS) break;
+                    }
+                    else if (c <= ' ')
+                    {
+                        hadWS = true;
```

Comment:
In which case is a space character usefully part of an expression?

Reply:
`test some` -> break before `s`
`test           ()` -> break after `()`

Reply:
We can not know in advance at what distance from the text the opening bracket will be and whether it will be at all.